google_guest_agent/instance_setup.go

```diff
@@ -193,6 +241,307 @@ func agentInit(ctx context.Context) {
 	}
 }
 
+func setMultiQueue(totalCPUs int) error {
+	devices, err := filepath.Glob(virtioNetDevs)
+	if err != nil {
+		return err
+	}
+	for _, dev := range devices {
+		if err := enableMultiQueue(dev); err != nil {
+			logger.Warningf("Could not enable multi queue for %s.", dev)
+			return err
+		}
+		if err = setQueueNumForDevice(dev); err != nil {
+			logger.Warningf("Could not set queue num for %s.", dev)
+			return err
+		}
+	}
+	// Set smp_affinity properly for gvnic queues. '-ntfy-block.' is unique to gve and will not affect virtio queues.
+	if err = setSMPAffinityForGVNIC(totalCPUs); err != nil {
+		logger.Warningf("Could not set smp_affinity for gvnic.")
+		return err
+	}
+	return nil
+}
+
+func setSMPAffinityForGVNIC(totalCPUs int) error {
+	irqDirs, err := filepath.Glob(irqDirPath)
+	if err != nil {
+		return err
+	}
+	for _, irq := range irqDirs {
+		blocks, err := filepath.Glob(irq + "/*-ntfy-block.*")
+		if err != nil {
+			return err
+		}
+		if len(blocks) > 0 && isFile(irq+"/affinity_hint") {
+			if err = copyFile(irq+"/affinity_hint", irq+"/smp_affinity"); err != nil {
+				return err
+			}
+		}
+	}
+	XPS, err := filepath.Glob(xpsCPU)
+	if err != nil {
+		return err
+	}
+
+	// If we have more CPUs than queues, then stripe CPUs across tx affinity as CPUNumber % queue_count.
+	for _, q := range XPS {
+		numQueues := len(XPS)
+		if numQueues > 63 {
+			numQueues = 63
+		}
+		r, _ := regexp.Compile(queueRegex)
+		match := r.MatchString(q)
+		var queueNum int
+		if match {
+			queueNum, err = strconv.Atoi(r.FindAllStringSubmatch(q, -1)[0][1])
+			if err != nil {
+				return err
+			}
+		}
+		xps := 0
+		for _, cpu := range makeRange(queueNum, numQueues, totalCPUs-1) {
```

Comment:
go has a c-like for syntax you can use here instead of defining the range function. `ForClause = [ InitStmt ] ";" [ Condition ] ";" [ PostStmt ] .`